core/Test/Tasty/Patterns.hs

```diff
@@ -44,7 +45,11 @@ instance IsOption TestPattern where
   parseValue = parseTestPattern
   optionName = return "pattern"
   optionHelp = return "Select only tests which satisfy a pattern or awk expression"
-  optionCLParser = mkOptionCLParser (short 'p' <> metavar "PATTERN")
+  optionCLParser = fmap (foldl' testPatternAnd noPattern) $ some $ mkOptionCLParser (short 'p' <> metavar "PATTERN")
+    where
+    testPatternAnd (TestPattern Nothing) (TestPattern mbY) = TestPattern mbY
+    testPatternAnd (TestPattern mbX) (TestPattern Nothing) = TestPattern mbX
+    testPatternAnd (TestPattern (Just x)) (TestPattern (Just y)) = TestPattern (Just (And x y))
```

Comment:
It feels that the `Nothing` cases should be impossible.
However, there is the weird behavior `parseTestPattern "" = Just noPattern`.
https://github.com/UnkindPartition/tasty/blob/c2cefb1839ead2713cf40a67463a0e4a0ea24e08/core/Test/Tasty/Patterns.hs#L57-L60
I would have expected `parseTestPattern "" = Nothing` (or even `parseTestPattern "" = undefined`), because
> The option `-p` expects an argument.

I would argue that the impossible cases should be marked in the source code as such,
```haskell
testPatternAnd (TestPattern (Just x)) (TestPattern (Just y)) = TestPattern (Just (And x y))
testPatternAnd _ _ = undefined
```
and covered by tests.

Reply:
`mkOptionCLParser` is too restrictive, but it should be possible to avoid concatenation of `Maybe Expr` entirely with something like 
```haskell
optionCLParser = optional $ fmap (foldl1' testPatternAnd) $ some $ bla-bla-bla (short 'p' <> metavar "PATTERN")
```
for a certain `bla-bla-bla` ;)

Reply:
I don't think that's possible, if you intend to preserve the `parseTestPattern "" = Just noPattern` behavior. The parse action can return a value or raise an error; if the string it receives is empty, it can't return a `Nothing` (or we're stuck catenating `Maybe`s), it can't return a `TestPattern` (because that wraps `Maybe`s), and it can't return an `Expr` (unless you want me to stuff the empty string into an `ERE` and then test it outside before `And`ing it to another `Expr`, which is just like catenating `Maybe`s but worse). What's left for it to return in that case?